flink-streaming-connectors/flink-connector-filesystem/src/main/java/org/apache/flink/streaming/connectors/fs/bucketing/BucketingSink.java

```diff
@@ -570,284 +583,278 @@ private void closeCurrentPartFile(BucketState<T> bucketState) throws Exception {
-								Path pendingPath = new Path(finalPath.getParent(),
-									pendingPrefix + finalPath.getName()).suffix(pendingSuffix);
+								Path pendingPath = getPendingPathFor(finalPath);
 
 								fs.rename(pendingPath, finalPath);
 								LOG.debug(
 									"Moving pending file {} to final location having completed checkpoint {}.",
 									pendingPath,
 									pastCheckpointId);
 							}
-							checkpointsToRemove.add(pastCheckpointId);
+							pendingIt.remove();
 						}
 					}
-					if (!bucketState.isWriterOpen && bucketState.pendingFiles.isEmpty()) {
+
+					if (!bucketState.isWriterOpen &&
+						bucketState.pendingFiles.isEmpty() &&
+						bucketState.pendingFilesPerCheckpoint.isEmpty()) {
+
 						// We've dealt with all the pending files and the writer for this bucket is not currently open.
 						// Therefore this bucket is currently inactive and we can remove it from our state.
-						it.remove();
-					} else {
-						for (Long toRemove : checkpointsToRemove) {
-							bucketState.pendingFilesPerCheckpoint.remove(toRemove);
-						}
+						stateIt.remove();
 					}
 				}
 			}
 		}
 	}
 
 	@Override
-	public State<T> snapshotState(long checkpointId, long checkpointTimestamp) throws Exception {
+	public void snapshotState(FunctionSnapshotContext context) throws Exception {
+		Preconditions.checkNotNull(restoredBucketStates, "The operator has not been properly initialized.");
+
+		restoredBucketStates.clear();
+
 		synchronized (state.bucketStates) {
-			for (BucketState<T> bucketState : state.bucketStates.values()) {
+			int subtaskIdx = getRuntimeContext().getIndexOfThisSubtask();
+
+			for (Map.Entry<String, BucketState<T>> bucketStateEntry : state.bucketStates.entrySet()) {
+				BucketState<T> bucketState = bucketStateEntry.getValue();
+
 				if (bucketState.isWriterOpen) {
-					long pos = bucketState.writer.flush();
-					bucketState.currentFileValidLength = pos;
+					bucketState.currentFileValidLength = bucketState.writer.flush();
 				}
+
 				synchronized (bucketState.pendingFilesPerCheckpoint) {
-					bucketState.pendingFilesPerCheckpoint.put(checkpointId, bucketState.pendingFiles);
+					bucketState.pendingFilesPerCheckpoint.put(context.getCheckpointId(), bucketState.pendingFiles);
 				}
 				bucketState.pendingFiles = new ArrayList<>();
 			}
+			restoredBucketStates.add(state);
+
+			if (LOG.isDebugEnabled()) {
+				LOG.debug("BucketingSink idx {} checkpointed {}.", subtaskIdx, state);
+			}
 		}
-		return state;
 	}
 
-	@Override
-	public void restoreState(State<T> state) {
-		this.state = state;
-
-		try {
-			initFileSystem();
-		} catch (IOException e) {
-			LOG.error("Error while creating FileSystem in checkpoint restore.", e);
-			throw new RuntimeException("Error while creating FileSystem in checkpoint restore.", e);
-		}
+	private void handleRestoredBucketState(BucketState<T> bucketState) {
+		// we can clean all the pending files since they were renamed to
```

Comment:
I don't understand why the subtask index even matters. In a simplified view, every subtask maintains a set of `File`s. When rescaling/restoring state, this set of files is given to another subtask. If this set contains a file marked as pending, it should be allowed to delete it, since no other task should have the same file since the state containing the information about the file (aka, the file path) is not given to multiple subtasks.